### PR TITLE
Update commons-beanutils to 1.9.4, adapt Stapler

### PR DESCRIPTION
**Pull request by daniel-beck**
Downstream from https://github.com/stapler/stapler/pull/211

Previously: #5319, #5246, #5124, #4928, #4328

This looks a lot like #5246 but has a different approach in Stapler: Rather than restore the previous permissive `beanutils` behavior (~ disabling the security fix), this attempts to patch Stapler so it doesn't require that workaround. The upside is we're doing more than just shutting up scanners, the downside is that we're doing more than just shutting up scanners 😅 

The fix in Stapler may be incomplete (it assumes the only custom tag with `class` attribute is `st:include` which seems very optimistic), so it probably makes sense to add an escape hatch just in case.

See https://github.com/jenkinsci/jenkins/pull/5324#issuecomment-802656978 for a summary of findings through tests and a lot of stepping through Jelly code.

### Proposed changelog entries

* Update commons-beanutils library to 1.9.4.
* Update Stapler to 1.263 to make the commons-beanutils update work.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by jglick** [comment]
Delete https://github.com/jenkinsci/jenkins/blob/39c33b8c998452241426cc2df181edbb490d7503/.github/dependabot.yml#L17-L19

**Comment by basil** [comment]
> It seems in my (fairly limited so far) testing that the problem is `st:include`'s `class` attribute (different from Jelly's tags, which mostly use `className` instead).

Good point.

> So this pull request renames the attribute to `className` (and keeps the existing setter for compatibility).

I agree this is the better approach.

> The fix in Stapler may be incomplete (it assumes the only custom tag with `class` attribute is `st:include` which seems very optimistic)

We have a rich body of data available in `usage-in-plugins`, so I did some digging to see whether or not this assumption is safe. First, I wrote some throwaway spaghetti code in `usage-in-plugins` to parse all Jelly files in Jenkins core and all plugins (including CloudBees proprietary plugins). Then I printed the values of the `class` attribute on all `<st:include/>` tags. I looked at the results and came up with a heuristic to classify the `class` attributes on other tags as either Java classes or CSS classes. The heuristic looked for e.g. FQCNs that start with `hudson.` or `jenkins.`, Groovy string interpolation of fields named `clazz`, etc. The heuristic matched 97% of the `class` attributes on `<st:include/>` tags, so the goal was to see if it would flag any `class` attributes on other tags. Well, after sifting through 20,632 `class` attributes on other tags, it eliminated 99% of them as obviously not Java classes and left me with 123 cases to study. Examining the remaining `class` attributes manually, it was very clear that these were all false positives; i.e., CSS class attributes and not Java class names.

So yes, all `class` attributes that refer to Java classes are on `<st:include/>` tags. For reference, here is the full list of tags that contain `class` attributes. For the `<st:include/>` tags these `class` attributes mean "Java class", while for all the others, they mean "CSS class":

 - `b:alert`
 - `bs:card`
 - `cbcipm:tabcontent`
 - `dt:blockWrapper`
 - `f:checkbox`
 - `f:entry`
 - `f:number`
 - `f:password`
 - `f:radio`
 - `f:repeatableDeleteButton`
 - `f:section`
 - `f:select`
 - `f:textarea`
 - `f:textbox`
 - `fa:svg-icon`
 - `l:confirmationLink`
 - `l:copyButton`
 - `l:helpIcon`
 - `l:icon`
 - `l:pane`
 - `l:svgIcon`
 - `lsi:icon`
 - `m:div`
 - `m:input`
 - `m:select`
 - `poj:textarea`
 - `s:include`
 - `sf:select`
 - `st:include`

Based on this I agree we should move forward with changing `<st:include class=""/>` to `<st:include className=""/>`. So just how big of a migration are we dealing with here? Across the archives I scanned (which includes both open source and CloudBees proprietary plugins) I found 180 plugins that needed updating (as well as Jenkins core). There were 439 unique Jelly files that needed updating. There were 1,070 unique `<std:include class=""/>` attributes that needed updating.

I wonder if it might be feasible to script the transformation and apply it at least to Jenkins core as part of the PR that upgrades Stapler. As for the other 180 plugins, maybe we resurrect @daniel-beck-bot?

**Comment by basil** [comment]
> I wonder if it might be feasible to script the transformation and apply it at least to Jenkins core as part of the PR that upgrades Stapler

I wrote up a quick and dirty script to do so at https://github.com/basil/jelly-migrator/ and filed #5329 to demonstrate the result of the automated migration. So if we want to begin this migration, that would be one repository down and 180 more to go.

**Comment by daniel-beck** [comment]
@basil 

> So just how big of a migration are we dealing with here?

Ideally none, as this should already be taken care of Stapler by https://github.com/stapler/stapler/pull/211/files#diff-98be0c72324bbd3fe59fa76673ad243e208349cda0985231aa5171021bd8c776R109-R116

**Comment by daniel-beck** [comment]
> So yes, all `class` attributes that refer to Java classes are on `<st:include/>` tags. For reference, here is the full list of tags that contain `class` attributes. For the `<st:include/>` tags these `class` attributes mean "Java class", while for all the others, they mean "CSS class":

Quite a few of these are probably bugs, see https://github.com/jenkinsci/jenkins/pull/5325

I'm also not quite sure yet what's special about `st:include` that makes it fail here when nothing else does. Might be that the type matters and not just the name, might be that it's an explicit attribute rather than a `ReallyStaticTagLibrary` or morph library passthrough, …

**Comment by basil** [comment]
> Ideally none

Not sure I follow the reasoning here. If the `class` attribute is deprecated in favor of `clazz`, then this implies that new code ought to use the non-deprecated syntax and old code, while still working, ought to be _eventually_ converted from the deprecated syntax to the non-deprecated syntax. If the migration from deprecated to non-deprecated syntax is not completed in a timely fashion, the net result is an increase in the complexity and cognitive load of the system (which inhibits its long-term sustainability). For this reason, static analyzers flag usages of deprecated APIs as a code smell.

If we do not intend to eventually convert consumers from the old syntax to the new syntax, I would suggest not introducing the new syntax but rather preserving the existing syntax in the public API (i.e., only supporting `class`), with the conversion from `class` to `clazz` being merely an internal implementation detail within Stapler.

**Comment by daniel-beck** [comment]
> then this implies that new code ought to use the non-deprecated syntax and old code, while still working, ought to be _eventually_ converted from the deprecated syntax to the non-deprecated syntax

I agree, _eventually_. Way down the priority list here IMO, given we may completely blow up some plugins with the current approach. And not migrating core immediately might actually give us a better chance to catch some edge cases as it's basically validating the compatibility code.

**Comment by basil** [comment]
> Way down the priority list here IMO

Perhaps it might be worth considering whether this project has a habit of starting [migrations](https://umach.medium.com/managing-migrations-1312a5e3ac9f) that never get completed and whether the resulting increase in complexity and cognitive load is hurting the project. With 1,389 usages of deprecated APIs within Jenkins core itself (and many more in plugins), historical trends show that "way down the priority list" may very well turn into "never". I have dedicated a significant portion of my career to reducing complexity and cognitive load by either completing half-finished migrations or else rolling them back. My general observation is that without having the organizational structure in place to effectively execute migrations (as described in Uma's article), it is better to avoid starting them in the first place rather than to start them and then not complete them.

**Comment by daniel-beck** [comment]
> historical trends show that "way down the priority list" may very well turn into "never".

That's valid, although I wrote that originally because I am not even confident this approach will reliably work, so adapting views may well turn out to be wasted time.

That said, I like your suggestion of continuing to use `class` in views; the wrapper I wrote in Stapler could easily be extended to throw on attempts to set `clazz`, and then it's purely implementation detail in Stapler, at least for this tag. If we don't find anything else that's broken, that seems like a viable approach not requiring downstream adaptations (or me to figure out how to even update the seemingly generated docs of Stapler).

**Comment by jglick** [comment]
> update the seemingly generated docs of Stapler

Evidently not any more; see INFRA-2908.

**Comment by daniel-beck** [comment]
> That said, I like your suggestion of continuing to use `class` in views; the wrapper I wrote in Stapler could easily be extended to throw on attempts to set `clazz`, and then it's purely implementation detail in Stapler, at least for this tag.

This is now implemented.

**Comment by daniel-beck** [comment]
AFAICT so far, the problem is

1. Custom Jelly tags …
2. … that do not implement `DynaTag` …
3. … with an attribute of any type called `class` …
4. … when invoked from Groovy views/fragments.

TODO next in addition to the test I now added:

- [ ] Add a test with a `CustomTagLibrary` to make sure they're not affected (as they're probably fairly common)
- [ ] Add a test with a `DefineTagLibrary` (i.e. local tags)
- [ ] Add a test with `ReallyStaticTagLibrary` (i.e. tags without namespace, basically direct HTML output)

**Comment by daniel-beck** [comment]
I'm running out of ideas what to test.

I think the last few test cases in `testUserDefinedTagLibrary` come closest to capturing the problem with `st:include`, but I would need to look closer. What I know so far and confirm through tests:

* Regular HTML is unaffected (duh).
* Custom tag libraries are unaffected (`/lib/form` etc.).
* Local tag libraries (using `jelly:define`'s `d:taglib` etc.) are unaffected.
* Jelly _views_ (as callers) are unaffected, only Groovy views are.

I also looked over the default Jelly tag suite and couldn't find anything with a `class` attribute. Given the hoops I had to jump through to have a custom tag library, and then it only works as expected in Groovy (which doesn't really need it), the likelihood of anyone having this problem is fairly low. And if they do, they can either:

* Implement `DynaTag`
* Apply the same taglib-level workaround I add in Stapler's `StaplerTagLibrary` for `st:include`.

**Comment by daniel-beck** [comment]
@basil Does https://github.com/jenkinsci/jenkins/pull/5324#issuecomment-794051449 implemented in https://github.com/stapler/stapler/pull/211/files#diff-e83c574d7de79fd9e24e450434121a8dbee2b436a19f103b32c936406dbf26ecR44-R47 look reasonable?

**Review by jeffret-b — APPROVED** [review]
This is basically the integration of the stapler change, which I and others have already approved. It's great to see a nice set of tests included in this.

We can't afford to get stuck behind on this upgrade and this is the best proposal we've had. It looks like it should work fine. If users do find problems in some scenarios, there are enough configuration options / escape hatches to allow them to continue while we fix any odd, undiscovered issues.

**Review by basil — APPROVED** [review]
This looks great! Very nice change.

**Review by oleg-nenashev — APPROVED** [review]
I do not see anything wrong in this pull request. Thanks to everyone for the analysis and tests. It is still a leap of faith to some extent, but it is a good timing to do such experiments. We can rollback weekly in the worst case.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by oleg-nenashev** [comment]
Does it make sense to add a developer changelog entry about deprecating the 'class' attribute names ?
AFAICT we would like developers to gradually migrate their taglibs to `clazz`, and to do the same in our main lib

**Comment by daniel-beck** [comment]
> Does it make sense to add a developer changelog entry about deprecating the 'class' attribute names ?
> AFAICT we would like developers to gradually migrate their taglibs to `clazz`, and to do the same in our main lib

Not really. It's a fine attribute in most cases, and I see no need for us to spam everyone with `st:include` updates (hence blocking the internal-only `clazz` attribute in Stapler).

The `class` attribute is no longer supported in very narrow, unusual use cases, as the tests demonstrate. Almost all tags are defined using `ReallyStaticTagLibrary` (for a loose definition of "defined"), `CustomTagLibrary` (with the `taglib` marker file) or using `jelly:define`, and none of those are affected.

**Comment by oleg-nenashev** [comment]
Ok, thanks! Maybe it worth to add a small note about the class attribute
support risks into the Jenkins developer guide, but I agree that there is
no massive action needed

**Comment by daniel-beck** [comment]
> risks into the Jenkins developer guide

FWIW we have no tests asserting the `st:include` problem case generically, because I couldn't figure out to inject a code-defined custom tag library in a way a plugin could. If you know how, please let me know so we can have better tests.

**Review by alecharp — APPROVED** [review]

**Review by olamy — APPROVED** [review]

File: bom/pom.xml
Comment (res0nance) [inline bom/pom.xml:44]: Do we need this released before merging?

Comment (daniel-beck) [reply bom/pom.xml:44]: Yes.

**Comment by jglick** [comment]
I would recommend marking as draft until you have a final dep version.

File: bom/pom.xml
Comment (basil) [inline bom/pom.xml:44]:
```suggestion
    <stapler.version>1.263</stapler.version>
```

**Review by basil — APPROVED** [review]

**Comment by MRamonLeon** [comment]
I think this is ready to merge if nobody objects.

We may merge it in 24 hours if there is no negative feedback. Please see the merge process documentation for more information about the merge process

**Pull request merged**
